node/src/components/network/limiter.rs

```diff
@@ -112,15 +113,15 @@ impl Limiter {
     }
 
     pub(super) fn debug_inspect_unspent_allowance(&self) -> Option<i64> {
-        // Disabled temporarily, as it results in a panic:
-
-        // node panicked: Cannot start a runtime from within a runtime. This happens because a
-        // function (like `block_on`) attempted to block the current thread while the thread is
-        // being used to drive asynchronous tasks.
-
-        None
-
-        //Some(self.data.resources.blocking_lock().available)
+        // Calling `blocking_lock()` from withing an async runtime will cause Tokio to panic with
+        // "Cannot start a runtime from within a runtime..." message. To avoid that we do the
+        // blocking call in a dedicated thread. It's a workaround that has a slight performance
+        // cost, but it's only invoked when using the `net-info` command of the Diagnostics Port,
+        // hence it's assumed to be rare.
+        let resources = Arc::clone(&self.data.resources);
+        thread::spawn(move || Some(resources.blocking_lock().available))
```

Comment:
The tokio devs also warn about this construct and mention that the "correct hack" here is to use `block_in_place`  instead of spawning a thread, which is less prone to accidental deadlocks. Let's change it!

Reply:
To give a little more detail: A deadlock would occur in the following conditions:

1. The lock is held by an async task that then goes to sleep (`tokio::time::sleep`) to wait for enough resources to pile up. It returns and is scheduled behind the current `handle_event`, which is already running.
2. `handle_event` calls the `debug_inspect_unspent_allowance`, which spawns a thread that wants to acquire the lock.
3. The thread is sleeping, waiting for the lock to be released.
4. Meanwhile, `debug_inspect_unspent_allowance` also cannot make any progress, since it is still running the `join()` call. This call will never finish, since on the same thread, the task that would release it is scheduled (and will never run).

This will then deadlock the entire node, as that is the main event processing thread.

Reply:
In contrast, `block_in_place` solves this situation. From its documentation:

> Calling this function informs the executor that the currently executing task is about to block the thread, so the executor is able to hand off any other tasks it has to a new worker thread before that happens.

Specifically, the task that will ultimately release the lock (see 4. above) will be moved to a different async worker thread and run eventually.

Reply:
TIL :+1: 
Thanks for digging into details!

Change to `block_in_place` is here: https://github.com/casper-network/casper-node/pull/3389/commits/a7bfbc96595526db80f7cc5d0b76958740bdd11c